From e6045d2e8021a414eded280701cd94cdf0b2be6d Mon Sep 17 00:00:00 2001
From: Suresh Kumar Gautam <sgkuma@codeaurora.org>
Date: Mon, 6 Nov 2017 15:25:36 +0530
Subject: [PATCH] Increased delay for disconnecting client to 1 sec,as 10ms is
 too aggressive for stations. Cancel disconnect timer from Deauth/Disassoc
 Event handler.

---
 src/ap/ieee802_1x.c |  2 +-
 src/ap/sta_info.c   | 22 ++++++++++++++++++++--
 2 files changed, 21 insertions(+), 3 deletions(-)

diff --git a/src/ap/ieee802_1x.c b/src/ap/ieee802_1x.c
index ae38f6c..7d7f27d 100644
--- a/src/ap/ieee802_1x.c
+++ b/src/ap/ieee802_1x.c
@@ -2370,7 +2370,7 @@ int ieee802_1x_eapol_tx_status(struct hostapd_data *hapd, struct sta_info *sta,
 	    ap_sta_pending_delayed_1x_auth_fail_disconnect(hapd, sta)) {
 		wpa_printf(MSG_DEBUG,
 			   "WPS: Indicate EAP completion on ACK for EAP-Failure");
-		hostapd_wps_eap_completed(hapd);
+		//hostapd_wps_eap_completed(hapd);
 	}
 #endif /* CONFIG_WPS */
 
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index ea217f3..4feab63 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -1296,6 +1296,15 @@ void ap_sta_deauth_cb(struct hostapd_data *hapd, struct sta_info *sta)
 	sta->flags &= ~WLAN_STA_PENDING_DEAUTH_CB;
 	eloop_cancel_timeout(ap_sta_deauth_cb_timeout, hapd, sta);
 	ap_sta_deauth_cb_timeout(hapd, sta);
+	if (eloop_cancel_timeout(ap_sta_delayed_1x_auth_fail_cb, hapd, sta) > 0)
+	{
+		wpa_printf(MSG_DEBUG,
+			   "%s: Removed ap_sta_delayed_1x_auth_fail_cb timeout for "
+			   MACSTR,
+			   hapd->conf->iface, MAC2STR(sta->addr));
+		if (sta->flags & WLAN_STA_WPS)
+			hostapd_wps_eap_completed(hapd);
+	}
 }
 
 
@@ -1308,6 +1317,15 @@ void ap_sta_disassoc_cb(struct hostapd_data *hapd, struct sta_info *sta)
 	sta->flags &= ~WLAN_STA_PENDING_DISASSOC_CB;
 	eloop_cancel_timeout(ap_sta_disassoc_cb_timeout, hapd, sta);
 	ap_sta_disassoc_cb_timeout(hapd, sta);
+	if (eloop_cancel_timeout(ap_sta_delayed_1x_auth_fail_cb, hapd, sta) > 0)
+	{
+		wpa_printf(MSG_DEBUG,
+			   "%s: Removed ap_sta_delayed_1x_auth_fail_cb timeout for "
+			   MACSTR,
+			   hapd->conf->iface, MAC2STR(sta->addr));
+		if (sta->flags & WLAN_STA_WPS)
+			hostapd_wps_eap_completed(hapd);
+	}
 }
 
 
@@ -1390,7 +1408,7 @@ void ap_sta_delayed_1x_auth_fail_disconnect(struct hostapd_data *hapd,
 {
 	wpa_dbg(hapd->msg_ctx, MSG_DEBUG,
 		"IEEE 802.1X: Force disconnection of " MACSTR
-		" after EAP-Failure in 10 ms", MAC2STR(sta->addr));
+		" after EAP-Failure in 1 sec", MAC2STR(sta->addr));
 
 	/*
 	 * Add a small sleep to increase likelihood of previously requested
@@ -1398,7 +1416,7 @@ void ap_sta_delayed_1x_auth_fail_disconnect(struct hostapd_data *hapd,
 	 * operations.
 	 */
 	eloop_cancel_timeout(ap_sta_delayed_1x_auth_fail_cb, hapd, sta);
-	eloop_register_timeout(0, 10000, ap_sta_delayed_1x_auth_fail_cb,
+	eloop_register_timeout(1, 10000, ap_sta_delayed_1x_auth_fail_cb,
 			       hapd, sta);
 }
 
-- 
1.9.1

